src/mame/drivers/magicard.cpp

```diff
@@ -350,313 +551,616 @@ TODO: check this register,doesn't seem to be 100% correct.
 	PORT_DIPSETTING(    0x80, DEF_STR( Off ) )
 	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
 
-	PORT_DIPNAME( 0x0100, 0x0100, "SYSTEM1" )
-	PORT_DIPSETTING(    0x0100, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	// Used by magicard to enter into gameplay
-	PORT_DIPNAME( 0x0200, 0x0200, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x0200, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	PORT_DIPNAME( 0x0400, 0x0400, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x0400, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	PORT_DIPNAME( 0x0800, 0x0800, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x0800, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	PORT_DIPNAME( 0x1000, 0x1000, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x1000, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	PORT_DIPNAME( 0x2000, 0x2000, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x2000, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	PORT_DIPNAME( 0x4000, 0x4000, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x4000, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
-	// coin sound in magicard (but no GFX is updated?)
-	PORT_DIPNAME( 0x8000, 0x8000, DEF_STR( Unknown ) )
-	PORT_DIPSETTING(    0x8000, DEF_STR( Off ) )
-	PORT_DIPSETTING(    0x0000, DEF_STR( On ) )
+	PORT_DIPNAME( 0x40, 0x40, "Hopper-Wert" )				PORT_DIPLOCATION("DIP 1:2")
+	PORT_DIPSETTING(    0x40, "10" )
+	PORT_DIPSETTING(    0x00, "5" )
+	
+	PORT_DIPNAME( 0x08, 0x08, "Munzer1" )					PORT_DIPLOCATION("DIP 1:5")
+	PORT_DIPSETTING(    0x08, "10" )
+	PORT_DIPSETTING(    0x00, "5" )
+	
+	PORT_DIPNAME( 0x04, 0x04, "Munzer2" )					PORT_DIPLOCATION("DIP 1:6")
+	PORT_DIPSETTING(    0x04, "10" )
+	PORT_DIPSETTING(    0x00, "5" )
+
+	PORT_DIPNAME( 0x02, 0x02, "Remote 1" )					PORT_DIPLOCATION("DIP 1:7")
+	PORT_DIPSETTING(    0x02, "100" )
+	PORT_DIPSETTING(    0x00, "10" )
+
+	PORT_DIPNAME( 0x20, 0x20, "Remote 2" )					PORT_DIPLOCATION("DIP 1:3")
+	PORT_DIPSETTING(    0x20, "100" )
+	PORT_DIPSETTING(    0x00, "50" )
+
+	PORT_DIPNAME( 0x10, 0x10, DEF_STR( Unknown ) )			PORT_DIPLOCATION("DIP 1:4")
+	PORT_DIPSETTING(    0x10, DEF_STR( Off ) )
+	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
+
+	PORT_DIPNAME( 0x01, 0x01, DEF_STR( Unknown ) )			PORT_DIPLOCATION("DIP 1:8")
+	PORT_DIPSETTING(    0x01, DEF_STR( Off ) )
+	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
+
+INPUT_PORTS_END
+
+static INPUT_PORTS_START( hotslots )
+	PORT_INCLUDE( magicard )
+
+	PORT_MODIFY("SW1")
+	PORT_BIT( 0x4000, IP_ACTIVE_LOW, IPT_BUTTON8 )     PORT_NAME("Alarm")           PORT_CODE(KEYCODE_S)
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_BUTTON9 )     PORT_NAME("Counter Control") PORT_CODE(KEYCODE_D)
+	PORT_BIT( 0x0400, IP_ACTIVE_LOW, IPT_BUTTON12 )    PORT_NAME("Clear coinCard")  PORT_CODE(KEYCODE_H)
+
 INPUT_PORTS_END
 
+void magicard_base_state::machine_reset()
+{
+	uint16_t *src = (uint16_t*)memregion("maincpu")->base();
+	m_scc66470->set_vectors( src );
```

Comment:
This seems like a gross way of doing this.  How does it actually work?  Does the 66470 have access to the CPU ROM?  Can it be configured as as a devcb or an address space?

Reply:
The 66470 provides DRAM at address 0x0-0x17ffff and a chip select for CPU ROM at 0x180000-0x1ffbff (so not the full size of the ROM).
On reset the CPU fetches PC and stack values from 0x0-0x7 which would normally be DRAM and therefore random data. So, for the first accesses to this area the 66470 asserts the ROM c/s instead of returning DRAM data. After this it'll return DRAM data for those locations.
The 66470 doesn't do anything with the ROM other than control its c/s line.
It also provides other chip selects,  the address of these are currently defined in the magicard.cpp map. Should this have been done in a different way, ie, declared in the 66470 map and some sort of callback ?  This chip doesn't do anything with the data and the data does not pass through it, it just provides the c/s lines.

Reply:
There definitely has to be a better way than this.  @galibert can you suggest something?